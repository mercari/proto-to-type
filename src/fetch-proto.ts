import protobuf from 'protobufjs';

export type ImportResolver = (target: string, weak: boolean) => ProtoRequest | null;
export interface FetcherOptions {
  /**
   * filter and transform import target string
   */
  resolveImport: ImportResolver;
  /**
   * fetch proto content with resolved location
   */
  fetchers?: Record<string, ProtoFetcher>;
}

export type ProtoFetcher = (...args: string[]) => Promise<string> | string;
type ExtarctRequests<T extends Record<string, ProtoFetcher>> =
  | [string, ...string[]]
  | {
      [K in keyof T]: ((type: K, ...args: Parameters<T[K]>) => 0) extends (...args: infer U) => 0 ? U : never;
    }[keyof T];
export type ProtoRequest = ExtarctRequests<typeof defaultFetchers>;

export const defaultFetchers = {
  'github-contents': (repo: string, path: string, commit: string) => {
    const { execSync } = require('child_process') as typeof import('child_process');
    const response = execSync(`hub api '/repos/${repo}/contents/${path}?ref=${commit}'`);
    return Buffer.from(JSON.parse(response.toString('utf-8')).content, 'base64').toString();
  },
  'read-file': (file: string) => {
    const { readFileSync } = require('fs') as typeof import('fs');
    return readFileSync(file).toString();
  },
};

const routeFetcher = ([type, ...args]: ProtoRequest, custom: Partial<Record<string, ProtoFetcher>>) => {
  const fetch = custom[type] ?? (defaultFetchers as typeof custom)[type];
  if (fetch) {
    return fetch(...args);
  } else {
    // eslint-disable-next-line no-console
    console.warn(`Unknown fetcher type ${type}: set custom fetcher`);
    return null;
  }
};

/**
 * fetch and assign proto file
 * It retuns loaded package name and its dependencies.
 * If no fetch option is assigned, it uses `hub` command.
 */
export const fetchProtoOnce = async (root: protobuf.Root, request: ProtoRequest, options: FetcherOptions) => {
  const { resolveImport } = options;
  try {
    const proto = await routeFetcher(request, options.fetchers || {});
    if (!proto) return [[]] as const;

    const meta = protobuf.parse(proto, root);

    const dependencies: (ProtoRequest | null)[] = [
      ...(meta.imports ?? []).map((target) => resolveImport(target, false)),
      ...(meta.weakImports ?? []).map((target) => resolveImport(target, true)),
    ];

    // incert '.' to align name format to parser.ReflectionObject.fullname
    const packageName = meta.package && `.${meta.package}`;

    return [dependencies.filter<ProtoRequest>(Boolean as any), packageName] as const;
  } catch (e) {
    // eslint-disable-next-line no-console
    console.log(e.stack);
    process.exit(1);
  }
};

/**
 * fetch needed proto files recursively
 */
export const fetchProtoAll = async (root: protobuf.Root, location: string, options: FetcherOptions) => {
  const mainLocation = options.resolveImport(location, false);
  if (!mainLocation) {
    // eslint-disable-next-line no-console
    console.log('Failed to resolve main location.');
    process.exit(1);
  }

  const [dependencies, mainPackage] = await fetchProtoOnce(root, mainLocation, options);
  if (!mainPackage) {
    // eslint-disable-next-line no-console
    console.log('No main package name found.');
    process.exit(1);
  }

  const queue = [...dependencies];
  const fetched = new Set<string>([location]);
  while (queue.length) {
    const dependency = queue.shift()!;
    const id = dependency.join('');
    if (fetched.has(id)) continue;

    const [next] = await fetchProtoOnce(root, dependency, options);
    fetched.add(id);
    queue.push(...next);
  }
  return mainPackage;
};
