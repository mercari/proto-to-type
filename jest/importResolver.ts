import { ImportResolver } from '../src';
import path from 'path';

// https://github.com/googleapis/googleapis/blob/50af0530730348f1e3697bf3c70261f7daaf2981/google/api/metric.proto
const GOOGLE_APIS_REPO = 'googleapis/googleapis';
const GOOGLE_APIS_COMMIT = '50af0530730348f1e3697bf3c70261f7daaf2981';
const METRIC = 'google/api/metric.proto';
const LABEL = 'google/api/label.proto';
const LAUNCH_STAGE = 'google/api/launch_stage.proto';

// https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/duration.proto
const GOOGLE_PROTOBUF_REPO = 'protocolbuffers/protobuf';
const GOOGLE_PROTOBUF_COMMIT = '176f7db11d8242b36a3ea6abb1cc436fca5bf75d';
const DURATION = 'google/protobuf/duration.proto';

export const protoFiles = { METRIC, LABEL, LAUNCH_STAGE, DURATION };

const fixtures = ['sample.proto', 'no-package.proto'];

export const importResolverForTesting: ImportResolver = (target) => {
  if (fixtures.includes(target)) {
    return ['read-file', path.resolve(__dirname, '../src/__fixtures__/', target)];
  } else if (target.startsWith('google/api/')) {
    return ['github-contents', GOOGLE_APIS_REPO, target, GOOGLE_APIS_COMMIT];
  } else if (target.startsWith('google/protobuf/')) {
    return ['github-contents', GOOGLE_PROTOBUF_REPO, `src/${target}`, GOOGLE_PROTOBUF_COMMIT];
  }
  return null;
};
