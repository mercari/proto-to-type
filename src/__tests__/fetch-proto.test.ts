import protobuf from 'protobufjs';
import { wrap } from 'jest-snapshot-serializer-raw';
import { fetchProtoOnce, fetchProtoAll, defaultFetchers } from '../fetch-proto';
import { importResolverForTesting as resolveImport, protoFiles } from '../../jest/importResolver';

const exitSymbol = Symbol();

const spyExit = jest.spyOn(process, 'exit');
spyExit.mockImplementation(() => {
  throw exitSymbol;
});

const { METRIC, LABEL, LAUNCH_STAGE, DURATION } = protoFiles;

describe('fetchProtoOnce', () => {
  const resolvedRequest = resolveImport(METRIC, false)!;
  it('correctly parses proto file', async () => {
    const root = new protobuf.Root();
    const [dependencies, packageName] = await fetchProtoOnce(root, resolvedRequest, { resolveImport });

    const expected = [LABEL, LAUNCH_STAGE, DURATION].map((target) => resolveImport(target, false));

    expect(packageName).toBe('.google.api');
    expect(dependencies).toEqual(expected);
    expect(wrap(JSON.stringify(root, null, '  '))).toMatchSnapshot();
  });

  it('supports custom fetcher', async () => {
    const root = new protobuf.Root();
    const customized = jest.fn(defaultFetchers['github-contents']);
    await fetchProtoOnce(root, resolvedRequest, {
      resolveImport,
      fetchers: { 'github-contents': customized },
    });
    expect(customized.mock.calls.length).toBe(1);
  });

  it('exits if something wrong happens', () => {
    const root = new protobuf.Root();
    const promise = fetchProtoOnce(root, resolvedRequest, {
      resolveImport,
      fetchers: {
        'github-contents': () => {
          throw new Error('Error for testing');
        },
      },
    });
    expect(promise).rejects.toBe(exitSymbol);
  });
});

describe('fetchProtoAll', () => {
  it('correctly parses proto files recursively', async () => {
    const root = new protobuf.Root();
    const mainPackage = await fetchProtoAll(root, METRIC, { resolveImport });

    expect(mainPackage).toBe('.google.api');
    expect(wrap(JSON.stringify(root, null, '  '))).toMatchSnapshot();
  });

  it('exits if main location resolving failed', () => {
    const root = new protobuf.Root();
    const promise = fetchProtoAll(root, '', { resolveImport });
    expect(promise).rejects.toBe(exitSymbol);
  });

  it('exits if main package name is missing', () => {
    const root = new protobuf.Root();
    const promise = fetchProtoAll(root, 'no-package.proto', { resolveImport });
    expect(promise).rejects.toBe(exitSymbol);
  });
});
