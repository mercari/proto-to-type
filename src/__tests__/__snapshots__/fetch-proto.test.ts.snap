// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchProtoAll correctly parses proto files recursively 1`] = `
{
  "nested": {
    "google": {
      "nested": {
        "api": {
          "options": {
            "go_package": "google.golang.org/genproto/googleapis/api;api",
            "java_multiple_files": true,
            "java_outer_classname": "LaunchStageProto",
            "java_package": "com.google.api",
            "objc_class_prefix": "GAPI",
            "cc_enable_arenas": true
          },
          "nested": {
            "MetricDescriptor": {
              "fields": {
                "name": {
                  "type": "string",
                  "id": 1
                },
                "type": {
                  "type": "string",
                  "id": 8
                },
                "labels": {
                  "rule": "repeated",
                  "type": "LabelDescriptor",
                  "id": 2
                },
                "metricKind": {
                  "type": "MetricKind",
                  "id": 3
                },
                "valueType": {
                  "type": "ValueType",
                  "id": 4
                },
                "unit": {
                  "type": "string",
                  "id": 5
                },
                "description": {
                  "type": "string",
                  "id": 6
                },
                "displayName": {
                  "type": "string",
                  "id": 7
                },
                "metadata": {
                  "type": "MetricDescriptorMetadata",
                  "id": 10
                },
                "launchStage": {
                  "type": "LaunchStage",
                  "id": 12
                }
              },
              "nested": {
                "MetricDescriptorMetadata": {
                  "fields": {
                    "launchStage": {
                      "type": "LaunchStage",
                      "id": 1,
                      "options": {
                        "deprecated": true
                      }
                    },
                    "samplePeriod": {
                      "type": "google.protobuf.Duration",
                      "id": 2
                    },
                    "ingestDelay": {
                      "type": "google.protobuf.Duration",
                      "id": 3
                    }
                  }
                },
                "MetricKind": {
                  "values": {
                    "METRIC_KIND_UNSPECIFIED": 0,
                    "GAUGE": 1,
                    "DELTA": 2,
                    "CUMULATIVE": 3
                  }
                },
                "ValueType": {
                  "values": {
                    "VALUE_TYPE_UNSPECIFIED": 0,
                    "BOOL": 1,
                    "INT64": 2,
                    "DOUBLE": 3,
                    "STRING": 4,
                    "DISTRIBUTION": 5,
                    "MONEY": 6
                  }
                }
              }
            },
            "Metric": {
              "fields": {
                "type": {
                  "type": "string",
                  "id": 3
                },
                "labels": {
                  "keyType": "string",
                  "type": "string",
                  "id": 2
                }
              }
            },
            "LabelDescriptor": {
              "fields": {
                "key": {
                  "type": "string",
                  "id": 1
                },
                "valueType": {
                  "type": "ValueType",
                  "id": 2
                },
                "description": {
                  "type": "string",
                  "id": 3
                }
              },
              "nested": {
                "ValueType": {
                  "values": {
                    "STRING": 0,
                    "BOOL": 1,
                    "INT64": 2
                  }
                }
              }
            },
            "LaunchStage": {
              "values": {
                "LAUNCH_STAGE_UNSPECIFIED": 0,
                "EARLY_ACCESS": 1,
                "ALPHA": 2,
                "BETA": 3,
                "GA": 4,
                "DEPRECATED": 5
              }
            }
          }
        },
        "protobuf": {
          "options": {
            "csharp_namespace": "Google.Protobuf.WellKnownTypes",
            "cc_enable_arenas": true,
            "go_package": "github.com/golang/protobuf/ptypes/duration",
            "java_package": "com.google.protobuf",
            "java_outer_classname": "DurationProto",
            "java_multiple_files": true,
            "objc_class_prefix": "GPB"
          },
          "nested": {
            "Duration": {
              "fields": {
                "seconds": {
                  "type": "int64",
                  "id": 1
                },
                "nanos": {
                  "type": "int32",
                  "id": 2
                }
              }
            }
          }
        }
      }
    }
  }
}
`;

exports[`fetchProtoOnce correctly parses proto file 1`] = `
{
  "nested": {
    "google": {
      "nested": {
        "api": {
          "options": {
            "go_package": "google.golang.org/genproto/googleapis/api/metric;metric",
            "java_multiple_files": true,
            "java_outer_classname": "MetricProto",
            "java_package": "com.google.api",
            "objc_class_prefix": "GAPI"
          },
          "nested": {
            "MetricDescriptor": {
              "fields": {
                "name": {
                  "type": "string",
                  "id": 1
                },
                "type": {
                  "type": "string",
                  "id": 8
                },
                "labels": {
                  "rule": "repeated",
                  "type": "LabelDescriptor",
                  "id": 2
                },
                "metricKind": {
                  "type": "MetricKind",
                  "id": 3
                },
                "valueType": {
                  "type": "ValueType",
                  "id": 4
                },
                "unit": {
                  "type": "string",
                  "id": 5
                },
                "description": {
                  "type": "string",
                  "id": 6
                },
                "displayName": {
                  "type": "string",
                  "id": 7
                },
                "metadata": {
                  "type": "MetricDescriptorMetadata",
                  "id": 10
                },
                "launchStage": {
                  "type": "LaunchStage",
                  "id": 12
                }
              },
              "nested": {
                "MetricDescriptorMetadata": {
                  "fields": {
                    "launchStage": {
                      "type": "LaunchStage",
                      "id": 1,
                      "options": {
                        "deprecated": true
                      }
                    },
                    "samplePeriod": {
                      "type": "google.protobuf.Duration",
                      "id": 2
                    },
                    "ingestDelay": {
                      "type": "google.protobuf.Duration",
                      "id": 3
                    }
                  }
                },
                "MetricKind": {
                  "values": {
                    "METRIC_KIND_UNSPECIFIED": 0,
                    "GAUGE": 1,
                    "DELTA": 2,
                    "CUMULATIVE": 3
                  }
                },
                "ValueType": {
                  "values": {
                    "VALUE_TYPE_UNSPECIFIED": 0,
                    "BOOL": 1,
                    "INT64": 2,
                    "DOUBLE": 3,
                    "STRING": 4,
                    "DISTRIBUTION": 5,
                    "MONEY": 6
                  }
                }
              }
            },
            "Metric": {
              "fields": {
                "type": {
                  "type": "string",
                  "id": 3
                },
                "labels": {
                  "keyType": "string",
                  "type": "string",
                  "id": 2
                }
              }
            }
          }
        }
      }
    }
  }
}
`;
